src/CheckinsManager.php

```diff
@@ -0,0 +1,172 @@
+            }
+            else {
+                $projectCheckins = $this->client->listForProject($localCheckin->projectId);
+                $filtered = array_filter($projectCheckins, function ($projectCheckin) use ($localCheckin) {
+                    return $projectCheckin->name === $localCheckin->name;
+                });
+                if (count($filtered) > 0) {
+                    $existingCheckin = $filtered[0];
+                    $localCheckin->id = $existingCheckin->id;
+                }
+            }
+
+            if ($existingCheckin) {
+                if (! $existingCheckin->isInSync($localCheckin)) {
+                    if ($updated = $this->update($localCheckin)) {
+                        $result[] = $updated;
+                    }
+                } else {
+                    // no change - just add to resulting array
+                    $result[] = $existingCheckin;
+                }
+            }
+            else if ($created = $this->create($localCheckin)) {
+                $result[] = $created;
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Loop through existing checkins and
+     * remove any that are not in the local checkins array.
+     *
+     * @param Checkin[] $localCheckins
+     * @return Checkin[]
+     */
+    private function removeNotFoundCheckins(array $localCheckins): array
+    {
+        $result = [];
+
+        $projectIds = array_unique(array_map(function ($checkin) {
+            return $checkin->projectId;
+        }, $localCheckins));
+
+        foreach ($projectIds as $projectId) {
+            $projectCheckins = $this->client->listForProject($projectId);
+            foreach ($projectCheckins as $projectCheckin) {
+                $filtered = array_filter($localCheckins, function ($checkin) use ($projectCheckin) {
+                    return $checkin->id === $projectCheckin->id;
+                });
+                if (count($filtered) === 0 && $this->remove($projectCheckin)) {
+                    $projectCheckin->markAsDeleted();
+                    $result[] = $projectCheckin;
+                }
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * @param Checkin $checkin
+     * @return Checkin|null
+     */
+    private function create(Checkin $checkin): ?Checkin
+    {
+        return $this->client->create($checkin);
+    }
+
+    /**
+     * @param Checkin $checkin
+     * @return Checkin|null
+     */
+    private function update(Checkin $checkin): ?Checkin
+    {
+        return $this->client->update($checkin);
+    }
+
+    /**
+     * @param Checkin $checkin
+     * @return bool
+     */
+    private function remove(Checkin $checkin): bool
+    {
+        return $this->client->remove($checkin->projectId, $checkin->id);
+    }
+
+    /**
+     * @param ServiceException $e
+     * @return void
+     */
+    private function handleServiceException(ServiceException $e): void
```

Comment:
Hmm, why is this method here and why does it manually invoke the service exception handler?

Reply:
This is here because the [`$checkin->validate()` method](https://github.com/honeybadger-io/honeybadger-php/pull/173/files#diff-677ca7bcc888f0ccacf5c99c0345aad1fc2abf4922e9942a00139128bb4ae834R43) can throw an error if the configuration is invalid. At first, I opted for throwing the exceptions and not allowing them to be configured by the user, but I remembered a discussion you and I had in the past: this being an exception reporting package, it should never throw unhandled exceptions :)

The implementation is the same as [`HoneybadgerClient.php`](https://github.com/honeybadger-io/honeybadger-php/blob/a81e96ceebd089ed33249dad7a81cd9a981385ed/src/HoneybadgerClient.php#L83). The default behaviour we set for the Laravel package is to [log a warning](https://github.com/honeybadger-io/honeybadger-laravel/pull/111/files#diff-74ffe2cea11eaa23d7fa43b69ba8489b02308ebff899f9bb63954311ba282c0cR192), same as the Honeybadger client.

Reply:
I'm not convinced that an error reporting package should never throw an error... when you set up a package for the first time it can be really helpful if it errors when you've screwed it up. So I think an error like "You forgot to pass in your API key" can be beneficial. 

But logging a warning could also be fine, I'm not particular on this 🤷‍♀️

Reply:
> I'm not convinced that an error reporting package should never throw an error

I think it depends on the path. As you said, in the setup path, it should throw an error. But not at runtime. If this class is meant to be invoked during deployments, though, then maybe you do want to throw an error.

Reply:
My original approach was that we do not want to block their deployment process. But this specific block is only to handle configuration validation. I guess I could change it to throw for this case.

Reply:
Yes. Also, since this is a lower-level implementation, which can be called from another PHP script or Laravel command, you should be bubbling up the original error to avoid information loss. The higher-level interface can then decide how to present that error.

Reply:
Done!